pkg/redis/redis.go

```diff
@@ -668,49 +682,183 @@ type Scripter interface {
+	ms := d.Milliseconds()
+	if ms == 0 && d > 0 {
+		return 1
+	}
+	return ms
+}
+
+// DeduplicateProtos deduplicates protos using key k. It stores a lock at LockKey(k) and the list of collected protos at ListKey(k).
 func DeduplicateProtos(ctx context.Context, r Scripter, k string, window time.Duration, msgs ...proto.Message) (bool, error) {
 	args := make([]interface{}, 0, 1+len(msgs))
-	args = append(args, window.Milliseconds())
+	args = append(args, milliseconds(window))
 	for _, msg := range msgs {
 		s, err := MarshalProto(msg)
 		if err != nil {
 			return false, err
 		}
 		args = append(args, s)
 	}
-	v, err := deduplicationScript.Run(r, []string{DeduplicationLockKey(k), DeduplicationListKey(k)}, args...).Result()
+	res, err := deduplicateProtosScript.Run(r, []string{LockKey(k), ListKey(k)}, args...).Int64()
 	if err != nil {
 		return false, ConvertError(err)
 	}
-	res, ok := v.(int64)
-	if !ok {
-		return false, errStore.New()
-	}
 	return res == 1, nil
 }
+
+// NOTE: Time stops in lua scripts and expired keys stay available.
+
+// lockMutexScript attempts to acquire mutex lock.
+// It returns 0 if lock is acquired or active locks TTL otherwise.
+var lockMutexScript = redis.NewScript(`local pttl = redis.call('pttl', KEYS[1])
+if pttl > 0 then
+	return pttl
+else
+	redis.call('del', KEYS[2])
+	redis.call('set', KEYS[1], ARGV[1], 'px', ARGV[2])
+	return 0
+end`)
+
+// takeMutexLockScript attempts to take over the lock from previous caller.
+// It returns 1 if lock is acquired and 0 otherwise
+var takeMutexLockScript = redis.NewScript(`if redis.call('get', KEYS[1]) == ARGV[1] then
+	redis.call('del', KEYS[2])
+	redis.call('set', KEYS[1], ARGV[3], 'px', ARGV[2])
+	return 1
+else
+	return 0
+end`)
+
+// unlockMutexLockScript unlocks the mutex lock.
+var unlockMutexScript = redis.NewScript(`if redis.call('get', KEYS[1]) == ARGV[1] then
+	redis.call('lpush', KEYS[2], ARGV[1])
+	redis.call('pexpire', KEYS[1], ARGV[2])
+	redis.call('pexpire', KEYS[2], ARGV[2])
+end
+return redis.status_reply('OK')`)
+
+// LockMutex locks the value stored at k with a mutex with identifier id.
+// It stores the lock at LockKey(k) and list at ListKey(k).
+func LockMutex(ctx context.Context, r redis.Cmdable, k, id string, expiration time.Duration) error {
+	var hasDeadline bool
+	dl, ok := ctx.Deadline()
+	if ok {
+		hasDeadline = !dl.IsZero()
+	}
+
+	lockKey := LockKey(k)
+	listKey := ListKey(k)
+	expMS := milliseconds(expiration)
+	for {
+		ttlMS, err := lockMutexScript.Run(r, []string{lockKey, listKey}, id, expMS).Int64()
+		if err != nil {
+			return ConvertError(err)
+		}
+		if ttlMS < 0 {
+			panic(fmt.Errorf("negative TTL returned: %d ms", ttlMS))
+		}
+		if ttlMS == 0 {
+			return nil
+		}
+
+		timeout := time.Duration(ttlMS) * time.Millisecond
+		if hasDeadline {
+			until := time.Until(dl)
+			if until < timeout {
+				timeout = until
+			}
+		}
+		if timeout < time.Second {
+			// Necessary until https://github.com/go-redis/redis/issues/1363 is resolved.
```

Comment:
This will be removed once https://github.com/go-redis/redis/pull/1365 is merged and, I suppose, we migrate to v8 of the library.